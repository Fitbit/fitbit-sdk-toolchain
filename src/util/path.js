export function dirname(path) {
  let dir = path.split('/').slice(0, -1).join('/');
  if (dir === '') dir = '/';
  return dir;
}

export function basename(path) {
  return path.replace(/\\/g, '/').replace(/.*\//, '');
}

export function split(path) {
  return [dirname(path), basename(path)];
}
