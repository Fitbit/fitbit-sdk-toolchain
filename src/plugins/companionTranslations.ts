import { basename } from 'path';
import { promisify } from 'util';

import { LanguageTable, MessageTable } from '@fitbit-sdk/companion-gettext';
import { default as _glob } from 'glob';
import humanizeList from 'humanize-list';
import pofile from 'pofile';
import { dataToEsm } from 'rollup-pluginutils';

import BuildError from '../util/BuildError';
import { validateLanguageTag, supportedTags } from '../languageTag';

const glob = promisify(_glob);
const loadPOFile = promisify(pofile.load);

async function loadTranslations(filePath: string) {
  const po = await loadPOFile(filePath);
  const messages: MessageTable = {};

  for (const { msgid, msgstr } of po.items) {
    if (msgstr.length > 1) {
      throw new BuildError(
        `msgid "${msgid}" in file "${filePath}" has multiple msgstr values. This is not supported.`,
      );
    }
    messages[msgid] = msgstr[0];
  }

  return messages;
}

export default function companionTranslations(
  globPattern: string,
  defaultLanguage: string,
) {
  return async () => {
    const languagePaths = new Map<string, string>();
    const translations: LanguageTable = {};

    for (const filePath of await glob(globPattern)) {
      const languageTag = basename(filePath, '.po');

      if (!validateLanguageTag(languageTag)) {
        throw new BuildError(
          `Translation file "${filePath}" has a bad name. Translation files must be named ${humanizeList(
            supportedTags.map((tag) => tag + '.po'),
            { conjunction: 'or' },
          )}.`,
        );
      }

      const existingTranslations = languagePaths.get(languageTag);

      if (existingTranslations) {
        throw new BuildError(
          `More than one translation file found for language ${languageTag}. Found "${existingTranslations}" and "${filePath}".`,
        );
      }

      languagePaths.set(languageTag, filePath);
      translations[languageTag] = await loadTranslations(filePath);
    }

    if (!translations.hasOwnProperty(defaultLanguage)) {
      throw new BuildError(
        `No translation file found for default language "${defaultLanguage}"`,
      );
    }

    return dataToEsm(translations, { namedExports: false });
  };
}
