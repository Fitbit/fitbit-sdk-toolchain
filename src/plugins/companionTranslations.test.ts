import { posix as path } from 'path';

import companionTranslations from './companionTranslations';
import {
  errorMessageSerializer,
  normalizeSlash,
} from '../jestSnapshotSerializers';

const basePath = normalizeSlash(
  path.join(__dirname, '__test__', 'companionTranslations'),
);

expect.addSnapshotSerializer(errorMessageSerializer);
expect.addSnapshotSerializer({
  test(val) {
    return val instanceof Error && val.message.includes(basePath);
  },

  print(val, serialize) {
    const err = val as Error;
    err.message = err.message.replace(basePath, '<base>');
    return serialize(err);
  },
});

function expectTranslations(translationsPath: string, defaultLanguage: string) {
  return expect(
    companionTranslations(
      path.join(basePath, translationsPath),
      defaultLanguage,
    )(),
  );
}

it('builds a table from all the available translations', () =>
  expectTranslations('good/**/*.po', 'fr-FR').resolves.toMatchSnapshot());

it('throws when multiple files map to the same language', () =>
  expectTranslations(
    'language-collision/**/*.po',
    'en-US',
  ).rejects.toMatchSnapshot());

it('throws when a translation file has multiple msgstr values for the same msgid', () =>
  expectTranslations(
    'multiple-msgstr/en-US.po',
    'en-US',
  ).rejects.toMatchSnapshot());

it('throws if the default language is not found', () =>
  expectTranslations('good/**/*.po', 'it-IT').rejects.toMatchSnapshot());

it.each(['.po', 'a.po', 'english.po', 'enus.po'])(
  'throws when encountering the badly named file %j',
  (name: string) =>
    expectTranslations(`bad-name/${name}`, 'en-US').rejects.toMatchSnapshot(),
);
