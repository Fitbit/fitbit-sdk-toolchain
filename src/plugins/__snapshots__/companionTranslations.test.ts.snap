// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builds a table from all the available translations 1`] = `
"export default {
	en: {
		foo: \\"English\\"
	},
	\\"fr-FR\\": {
		foo: \\"France français\\"
	},
	fr: {
		foo: \\"Français\\"
	},
	\\"en-US\\": {
		foo: \\"US English\\"
	},
	\\"fr-CA\\": {
		foo: \\"Français canadien\\"
	}
};"
`;

exports[`throws if the default locale is not found 1`] = `[Error: No translation file found for default locale "it"]`;

exports[`throws when a translation file has multiple msgstr values for the same msgid 1`] = `[Error: msgid "foo" in file "<base>/multiple-msgstr/en.po" has multiple msgstr values. This is not supported.]`;

exports[`throws when encountering the badly named file ".po" 1`] = `[Error: Translation file "<base>/bad-name/.po" has a bad name. Translation files must have names in the form ll-cc.po or ll.po (e.g. en-US.po)]`;

exports[`throws when encountering the badly named file "a.po" 1`] = `[Error: Translation file "<base>/bad-name/a.po" has a bad name. Translation files must have names in the form ll-cc.po or ll.po (e.g. en-US.po)]`;

exports[`throws when encountering the badly named file "english.po" 1`] = `[Error: Translation file "<base>/bad-name/english.po" has a bad name. Translation files must have names in the form ll-cc.po or ll.po (e.g. en-US.po)]`;

exports[`throws when encountering the badly named file "enus.po" 1`] = `[Error: Translation file "<base>/bad-name/enus.po" has a bad name. Translation files must have names in the form ll-cc.po or ll.po (e.g. en-US.po)]`;

exports[`throws when multiple files map to the same language 1`] = `[Error: More than one translation file found for language en-US. Found "<base>/language-collision/EN-us.po" and "<base>/language-collision/lang/en-US.po".]`;
